add_inventory/wizards/inventory_wizard.py

```diff
@@ -0,0 +1,169 @@
+
+        file_content = base64.b64decode(self.file)
+        file_stream = io.StringIO(file_content.decode("utf-8"))
+        csv_reader = csv.reader(file_stream, delimiter=',')
+
+        next(csv_reader, None)  
+        company_id = self.env.company.id 
+
+        if self.operation == 'add':
+            picking_type = self.env['stock.picking.type'].search([
+                ('code', '=', 'incoming'),
+                ('company_id', '=', company_id)
+            ], limit=1)
+            location_id = self.env.ref('stock.stock_location_suppliers').id  
+            location_dest_id = self.location_id.id  
+
+        else:  
+            picking_type = self.env['stock.picking.type'].search([
+                ('code', '=', 'outgoing'),
+                ('company_id', '=', company_id)
+            ], limit=1)
+            location_id = self.location_id.id  
+            location_dest_id = self.env.ref('stock.stock_location_customers').id  
+
+        if not picking_type:
+            raise ValidationError("No suitable picking type found.")
+
+        picking = self.env['stock.picking'].create({
+            'partner_id': self.env.user.company_id.partner_id.id,
+            'location_id': location_id,
+            'location_dest_id': location_dest_id,
+            'picking_type_id': picking_type.id,  
+            'state': 'draft',
+            'move_type': 'direct',
+            'company_id': company_id
+        })
+
+        for row in csv_reader:
+            product_name = row[0]
+            serial_number = str(row[1]) if row[1] is not None else ''
+            quantity = float(row[2])
+            uom = row[3]
+            hsn_code = row[4]
+
+            product = self._get_or_create_product(product_name, uom, hsn_code, quantity, serial_number)
+
+            if (product.tracking in ('serial', 'lot') and not serial_number) or (product.tracking == 'none' and serial_number):
+                continue 
+
+            if (product.tracking == 'serial' and quantity > 1) or (product.tracking == 'lot' and quantity <= 1):
+                continue
+
+            lot = self._get_or_create_lot(serial_number, product)
+
+            if self.operation == 'add':
+                stock_quant = self._get_stock_quant(product, self.location_id, lot, quantity, 'add')
+                if not product.l10n_in_hsn_code and str(hsn_code).isdigit() and len(str(hsn_code)) in (4, 6, 8):
+                    product.l10n_in_hsn_code = str(hsn_code)
+                if stock_quant: 
+                    self._add_stock_move(picking, product, quantity, lot, location_id, location_dest_id)
+
+            elif self.operation == 'remove':                
+                stock_quant = self._get_stock_quant(product, self.location_id, lot, quantity, 'remove')
+                if stock_quant:
+                    self._add_stock_move(picking, product, quantity, lot, location_id, location_dest_id)
+
+        picking.action_confirm()
+        picking.action_assign()
```

Comment:
I would call only action_confirm, which automatically creates the move lines by calling action_assign.